### PR TITLE
✨ Add optimization history to result and iteration column to parameter history 

### PR DESCRIPTION
**Pull request by s-weigand**
This change allows inspecting the stdout output from `scipy.optimize.least_squares` during optimization and saving the current iteration to the parameter history, as well as creating a `pd.DataFrame` for plotting and inspection.

The `OptimizationHistory` class is written in a way that allows using it like a view of `pd.DataFrame` e.g.

```python
result.optimization_history.cost.plot()
```

However, since the wrapping functionality is dynamic at runtime you lose help information and autocompleting from static analyzers, which why the underlying  `pd.DataFrame` is also exposed via the `data` property which gives you the static analyzer goodies.

```python
result.optimization_history.data.cost.plot()
```

### Change summary

- [✨ Added TeeContext](https://github.com/glotaran/pyglotaran/commit/3b5036fca27aacb70a70270844e2d1e68ef172f6)
- [🧹 Don't save index in when saving ParameterHistory to csv](https://github.com/glotaran/pyglotaran/commit/8015f02d5cf9cfac7d3559c9f1b69db6321b96b2)
- [✨ Added iteration to parameter history](https://github.com/glotaran/pyglotaran/commit/789583d0238f3ecddda0b1762ee6deb4af53c02f)
- [👌⌨️ Improved typing of glotaran.project.dataclass_helpers.fromdict](https://github.com/glotaran/pyglotaran/commit/676e3ed0b1af31c436edb1f86598388a86ce6c2a)
- [✨ Added OptimizationHistory to Result](https://github.com/glotaran/pyglotaran/commit/a92b3a49fb0c2b1586e1e2070067a0bcb2384ea8)
- [🧪 Added roundtrip test to result yml saving](https://github.com/glotaran/pyglotaran/commit/a00bdde7ef85856917e36ca6ab460b0b7b770f26)
- [📚 Added plot of optimization_history to quickstart](https://github.com/glotaran/pyglotaran/commit/62f6447f63de60f967067040513ec8b6797ba9ac)
- [🔧 Remove more metadata from notebooks with nbstripout](https://github.com/glotaran/pyglotaran/pull/1134/commits/0a63eb87f1df2243981a19bfc5c9e0491a478831)
- [👌 Disallow usage of 'iteration' as parameter label](https://github.com/glotaran/pyglotaran/pull/1134/commits/524325722d298ddc3050f239e079e1e048259ad3)
- [♻️ Made get_current_optimization_iteration a static method of Optimizer](https://github.com/glotaran/pyglotaran/pull/1134/commits/fadb8e43ab49d1614470794eafb8264f9b079353)
- [♻️ Moved optimization_history module from project to optimization pkg](https://github.com/glotaran/pyglotaran/pull/1134/commits/1545bf03a240ea72f39b05d8a2474080e01359a3)

Links to the changed sections

- [Plot of optimization history](https://pyglotaran--1134.org.readthedocs.build/en/1134/notebooks/quickstart/quickstart.html#Optimizing-data)

### Checklist

- [x] ✔️ Passing the tests (mandatory for all PR's)
- [x] 🚧 Added changes to changelog (mandatory for all PR's)
- [ ] 👌 Closes issue (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 🧪 Adds new tests for the feature (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 📚 Adds documentation of the feature

### Closes issues

Might allow #233  or one of the related issues to be closed

### CODE REVIEW

File: glotaran/utils/tee.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            Text written to buffer.
+        """
+        return self.buffer.getvalue()
+
+    def flush(self) -> None:
+        """Flush values in the buffer."""
+        self.buffer.flush()
+
+
+def get_current_optimization_iteration(optimize_stdout: str) -> int:
```
Comment (joernweissenborn) [inline glotaran/utils/tee.py:66]: This is specific to optimization implementation. Therefore it belongs to the optimizer class. It should be a class method of the Optimizer.

File: glotaran/project/optimization_history.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+"""Module containing the ``OptimizationHistory`` class."""
```
Comment (joernweissenborn) [inline glotaran/project/optimization_history.py:1]: Move this module to the optimization package.

**Review by joernweissenborn — CHANGES_REQUESTED** [review]

File: glotaran/utils/tee.py (hunk 1 above)
Comment (s-weigand) [reply glotaran/utils/tee.py]: Done in [fadb8e4](https://github.com/glotaran/pyglotaran/pull/1134/commits/fadb8e43ab49d1614470794eafb8264f9b079353)

File: glotaran/project/optimization_history.py (hunk 2 above)
Comment (s-weigand) [reply glotaran/project/optimization_history.py]: Done in [1545bf0](https://github.com/glotaran/pyglotaran/pull/1134/commits/1545bf03a240ea72f39b05d8a2474080e01359a3)

**Comment by s-weigand** [comment]
@jsnel I triggered a binder build of the PR branch so you don't need to wait to [try it out](https://mybinder.org/v2/gh/s-weigand/pyglotaran/optimization-history?urlpath=lab/tree/docs/source/notebooks).

**Review by joernweissenborn — DISMISSED** [review]
Lgtm

File: glotaran/parameter/parameter_group.py [hunk 3]
```diff
@@ -564,7 +564,7 @@ def set_from_history(self, history: ParameterHistory, index: int):
             The history index.
         """
         self.set_from_label_and_value_arrays(
-            history.parameter_labels, history.get_parameters(index)
+            history.parameter_labels[1:], history.get_parameters(index)[1:]
```
[the remark this answers is not in the document]
Comment (s-weigand) [reply glotaran/parameter/parameter_group.py]: Addressed in [46a0cfa](https://github.com/glotaran/pyglotaran/pull/1134/commits/46a0cfa125424ea7cf65870c22d64ad279e97f9c)

File: glotaran/utils/tee.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+"""Module containing ``tee`` like functionality."""
```
[the remark this answers is not in the document]
Comment (s-weigand) [reply glotaran/utils/tee.py]: Addressed in [9d18b21](https://github.com/glotaran/pyglotaran/pull/1134/commits/9d18b214d2d3ff4f98f086db1aa69b612ed7695c)

**Review by jsnel — APPROVED** [review]
Looks good to me now.

**Pull request merged**